startupscript/vertex-ai-user-managed-notebook/post-startup.sh

```diff
@@ -752,6 +754,27 @@ EOF
 chown ${LOGIN_USER}:${LOGIN_USER} "${USER_BASHRC}"
 chown ${LOGIN_USER}:${LOGIN_USER} "${USER_BASH_PROFILE}"
 
+####################################
+# Run a user provided startup script
+####################################
+
+# If the user has provided a startup script, run it after workbench setup,
+# but before restarting Jupyter and running tests.
+
+readonly USER_STARTUP_SCRIPT="$(get_metadata_value "instance/attributes/terra-user-startup-script")"
+if [[ -n "${USER_STARTUP_SCRIPT}" ]]; then
+  emit "Running user provided startup script..."
+  readonly USER_STARTUP_SCRIPT_FILE="${USER_TERRA_CONFIG_DIR}/user-startup-script.sh"
+
+  # Copy the user's startup script to the user's .terra directory
+  curl -L "${USER_STARTUP_SCRIPT}" >"${USER_STARTUP_SCRIPT_FILE}"
+
+  # Run the user's startup script as the root user so that they may install packages
+  emit "Running user provided startup script, output to ${USER_STARTUP_OUTPUT_FILE}..."
+  chmod +x "${USER_STARTUP_SCRIPT_FILE}"
+  bash -c "${USER_STARTUP_SCRIPT_FILE}" >${USER_STARTUP_OUTPUT_FILE} 2>&1
```

Comment:
Is there something intentional about using "bash -c" versus just:

```
"${USER_STARTUP_SCRIPT_FILE}" >${USER_STARTUP_OUTPUT_FILE} 2>&1
```

Also, we should add a comment about whether we are intentionally letting the user's startup script fail the post-startup script, along with the apparent expectation that users will debug this just like any other startup script failure.

It would also be helpful if we emitted the location that we are depositing their script. Thinking about my own steps in debugging startup scripts like this, I'm often doubtful about whether I remembered to re-push the latest changes to GCS or if I put the right path in the launch command. I then have to go scrounge around and read the comments above to remember that it was downloaded to ` /opt/c2d/post_start.sh`.

Reply:
Good points. Removed "bash -c". Added comment to head of script explaining user startup script behavior and error handling.

The output would look something like this. I wonder if its redundant for the user to emit the user script path, as the output already shows the location when the variable is set.
```
+ readonly USER_STARTUP_SCRIPT=gs://startupscript-terra-vdevel-beamish-okra-2163/user.sh
+ USER_STARTUP_SCRIPT=gs://startupscript-terra-vdevel-beamish-okra-2163/user.sh
+ [[ -n gs://startupscript-terra-vdevel-beamish-okra-2163/user.sh ]]
+ readonly USER_STARTUP_SCRIPT_FILE=/home/jupyter/.terra/user-startup-script.sh
+ USER_STARTUP_SCRIPT_FILE=/home/jupyter/.terra/user-startup-script.sh
+ emit 'Downloading user startup script to /home/jupyter/.terra/user-startup-script.sh...'
++ date '+%Y-%m-%d %H:%M:%S'
+ echo '2023-10-31 15:46:59 Downloading user startup script to /home/jupyter/.terra/user-startup-script.sh...'
2023-10-31 15:46:59 Downloading user startup script to /home/jupyter/.terra/user-startup-script.sh...
```